packages/frontend-v2/hooks/useLocalStorage.ts

```diff
@@ -1,33 +1,38 @@
-import { useState, Dispatch, SetStateAction } from "react";
+import { useState, Dispatch, useEffect } from "react";
 
-// NextJS executes server side code first, and then client side. Window object is only present client side,
-// so window could be undefined. This ensures the window object exists before we use it.
-const isWindow = typeof window !== "undefined";
+// Enum used for formik data, these entries can be deleted later.
+// These are here for testing purposes.
+export enum LocalStorageKey {
+  token = "token"
+}
 
+/**
+ * Using this hook allows for any component to access localstorage.
+ * Accessing the hook allows for the functionality of setting and getting a value in localstorage.
+ * @param key Key for the localstorage value.
+ * @param defaultValue Default value returned if the hook cannot access the value in localstorage.
+ */
 export function useLocalStorage<T>(
-  key: string,
-  initialValue: T
-): [T, Dispatch<SetStateAction<T>>] {
-  const [storedValue, setStoredValue] = useState<T>(() => {
+  key: LocalStorageKey,
+  defaultValue: T
+): [T, Dispatch<T>] {
+  const [storedValue, setStoredValue] = useState<T>(defaultValue);
+
+  useEffect(() => {
     try {
-      const item = isWindow && window.localStorage.getItem(key);
-      return item ? JSON.parse(item) : initialValue;
+      const item = window.localStorage.getItem(key);
+      setStoredValue(item ? JSON.parse(item) : defaultValue);
     } catch (error) {
       console.error(error);
-      return initialValue;
     }
-  });
+  }, [])
 
-  const setValue = (value: T | ((value: T) => T)) => {
+  const setValue = (value: T) => {
```

Comment:
so according to the rules of hooks, refs should be stable, meaning that re-renders should not cause things to change that do not need to change. however whenever this hook rerenders, this function will be re-evaluated, producing a new ref, and whatever things have this setter in their dependency array will re-calculate needlessly. so in theory, this should be memoized with `useCallback` but this is probably overkill, so feel free to ignore this.

relevant docs reference: https://reactjs.org/docs/hooks-reference.html#usecallback

the use case i'm referring to here is where it says:

> This is useful when passing callbacks to optimized child components that rely on reference equality to prevent unnecessary renders (e.g. shouldComponentUpdate).